### PR TITLE
Merge Search feature

### LINKED ISSUE #117: As a user, I want to be able to search for a specific anime by its name or genre

### PR DESCRIPTION
**Pull request by EyoWeiChin**
Add:
- SearchCommand to search for anime titles and genre by a search term
- SearchParser to parse parameters for search terms
- SearchParserTest
- SearchCommandTest

Update:
- EXPECTED.txt and input.txt to reflect changes
- UserGuide.md with the new Search feature
- BrowserParserTest to remove redundant code
- CommandParser fix for paramFieldCheck() to throw for whitespace-only parameter

closes #117

### CODE REVIEW

File: src/main/java/seedu/duke/command/SearchCommand.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+        switch (searchType) {
+        case 0:
+            searchForAnime(animeData);
+            break;
+        case 1:
+            searchForGenre(animeData);
+            break;
+        default:
+            LOGGER.log(Level.SEVERE, SEARCH_TYPE_INVALID_LOG);
```
Comment (ChanJianHao) [inline src/main/java/seedu/duke/command/SearchCommand.java:52]: Good use of logger!

File: src/main/java/seedu/duke/command/SearchCommand.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        for (Anime anime : animeData.getAnimeDataList()) {
+            if (anime.getAnimeName().toLowerCase().contains(searchTerm)) {
+                result += ID_HEADER + anime.getAnimeID() + ID_CLOSER + anime.getAnimeName() + System.lineSeparator();
+            }
+        }
+    }
+
+    public void setSearchTerm(String searchTerm) {
+        this.searchTerm = searchTerm.toLowerCase();
+        this.searchType = 0;
```
Comment (ChanJianHao) [inline src/main/java/seedu/duke/command/SearchCommand.java]: Perhaps a constant would provide better clarity?

File: src/main/java/seedu/duke/parser/SearchParser.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    public SearchCommand parse(String description) throws AniException {
+        String[] paramGiven = parameterSplitter(description);
+        paramIsSetCheck(paramGiven);
+        parameterParser(paramGiven);
+        return searchCommand;
+    }
+
+    public void parameterParser(String[] paramGiven) throws AniException {
+        for (String param : paramGiven) {
+            String[] paramParts = param.split(" ", 2);
```
Comment (ChanJianHao) [inline src/main/java/seedu/duke/parser/SearchParser.java]: Introducing a constant here would be good?

File: src/test/java/seedu/duke/command/SearchCommandTest.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        assertThrows(AniException.class, () -> {
+            testSearch.execute(animeData, storageManager, user);
+        });
+
+        testSearch.searchType = BELOW_RANGE;
+        assertThrows(AniException.class, () -> {
+            testSearch.execute(animeData, storageManager, user);
+        });
+    }
+}
```
Comment (ChanJianHao) [inline src/test/java/seedu/duke/command/SearchCommandTest.java]: Might want to add a new line to EOF here.

File: src/test/java/seedu/duke/parser/SearchParserTest.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        SearchParser testParse = new SearchParser();
+        assertThrows(AniException.class, () -> {
+            testParse.parse(INVALID_PARAMETERS_TEST1);
+        });
+
+        assertThrows(AniException.class, () -> {
+            testParse.parse(INVALID_PARAMETERS_TEST2);
+        });
+    }
+}
```
Comment (ChanJianHao) [inline src/test/java/seedu/duke/parser/SearchParserTest.java]: How about a new line here?

File: src/main/java/seedu/duke/command/SearchCommand.java (hunk 2 above)
Comment (EyoWeiChin) [reply src/main/java/seedu/duke/command/SearchCommand.java]: I have extracted the searchType as a constant!

File: src/main/java/seedu/duke/parser/SearchParser.java (hunk 3 above)
Comment (EyoWeiChin) [reply src/main/java/seedu/duke/parser/SearchParser.java]: I have extracted out the values as constants.

File: src/test/java/seedu/duke/command/SearchCommandTest.java (hunk 4 above)
Comment (EyoWeiChin) [reply src/test/java/seedu/duke/command/SearchCommandTest.java]: I have added a newline here.

File: src/test/java/seedu/duke/parser/SearchParserTest.java (hunk 5 above)
Comment (EyoWeiChin) [reply src/test/java/seedu/duke/parser/SearchParserTest.java]: I have added a newline here as well, Thank you!

File: docs/UserGuide.md [hunk 6]
```diff
@@ -264,12 +301,14 @@ Coming soon!
...
-| Switch user        | `switchuser -n <USERNAME>` |
-| Browse             | `browse -s [name/rating] -p <1-26> -o [asc/dsc]` |
-| Create watchlist   | `watchlist -n <WATCHLIST_NAME>` |
-| List all watchlist | `watchlist -l` |
-| Select watchlist   | `watchlist -s <WATCHLIST_INDEX>` |
-| Delete watchlist   | `watchlist -d <WATCHLIST_INDEX>` |
-| Add to watchlist   | `add -a <ANIME_NAME>` |
\ No newline at end of file
+| Help               | `help`                                                   |
+| add user           | `adduser -n <USERNAME> -dob <dd/MM/yyyy> -g <GENDER>`    |
```
Comment (OngDeZhi) [inline docs/UserGuide.md]: Capitalize the first letter, same for line 314.

**Review by OngDeZhi — APPROVED** [review]
Other than the user guide capitalization for the first letter, it looks good!

File: docs/UserGuide.md (hunk 6 above)
Comment (EyoWeiChin) [reply docs/UserGuide.md]: Got it! I have rectified the issue. Thank you

File: src/main/java/seedu/duke/command/SearchCommand.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    protected static final String SEARCHING_BY_GENRE = "Searching By Genre";
+    protected static final String SEARCHING_BY_ANIME_NAME = "Searching By Anime Name";
+    protected static final String SEARCH_TYPE_INVALID = "Something went wrong with search input";
+    protected static final String SEARCH_TYPE_INVALID_LOG = "Search Type has the wrong values.";
+    protected static final int SEARCH_BY_NAME = 0;
+    protected static final int SEARCH_BY_GENRE = 1;
+
+    private static final Logger LOGGER = getAniLogger(SearchCommand.class.getName());
+
+    protected String searchTerm;
```
Comment (n3wsoldier) [inline src/main/java/seedu/duke/command/SearchCommand.java]: I think for command, private will be the better access modifiers, as there will not be any subclass to SearchCommand.

Comment (EyoWeiChin) [reply src/main/java/seedu/duke/command/SearchCommand.java]: I have made the changes, in all affected areas. Thank you for pointing it out.

**Review by n3wsoldier — APPROVED** [review]
LGTM :) 👍

**Pull request merged**